File: Frontend/web/types/booking.ts
/**
 * Tipos TypeScript para el sistema de reservas Airbnb
 */

export type BookingStatus =
  | 'pending_confirmation'    // Esperando confirmación del propietario
  | 'confirmed'               // Confirmada, esperando pago de reserva
  | 'reservation_paid'        // 50% pagado (reserva confirmada)
  | 'checked_in'              // Check-in realizado
  | 'completed'               // Reserva completada (100% pagado)
  | 'cancelled_by_guest'      // Cancelada por huésped
  | 'cancelled_by_host'       // Cancelada por propietario
  | 'cancelled_no_payment'    // Cancelada por falta de pago
  | 'refunded'                // Reembolsada

export type BookingPaymentStatus =
  | 'pending'              // Pendiente de pago
  | 'processing'           // Procesando pago
  | 'completed'            // Pago exitoso
  | 'failed'               // Pago fallido
  | 'refunded'             // Reembolsado
  | 'partially_refunded'   // Parcialmente reembolsado

export type PaymentType =
  | 'reservation'  // Pago de reserva (50%)
  | 'checkin'      // Pago al check-in (50%)
  | 'full'         // Pago completo (100%)
  | 'refund'       // Reembolso

export interface Booking {
  id: string
  listingId: string
  listingCreatedAt: string
  guestUserId: string
  hostUserId: string
  
  // Fechas de la reserva
  checkInDate: string  // ISO date string
  checkOutDate: string // ISO date string
  nights: number
  
  // Precios (en PEN)
  pricePerNight: number
  totalPrice: number
  reservationAmount: number  // 50% del total
  checkinAmount: number      // 50% del total
  serviceFee: number
  cleaningFee: number
  
  // Estado
  status: BookingStatus
  
  // Información de huéspedes
  numberOfGuests: number
  
  // Comunicación
  guestMessage?: string
  hostResponse?: string
  cancellationReason?: string
  
  // Timestamps
  createdAt: string
  updatedAt: string
  confirmedAt?: string
  reservationPaidAt?: string
  checkedInAt?: string
  completedAt?: string
  cancelledAt?: string
  
  // Metadata
  metadata?: Record<string, any>
  
  // Relaciones (populadas)
  listing?: {
    id: string
    title: string
    images: string[]
    address: string
    district: string
  }
  guest?: {
    id: string
    firstName: string
    lastName: string
    email: string
    phone?: string
    profilePicture?: string
  }
  host?: {
    id: string
    firstName: string
    lastName: string
    email: string
    phone?: string
    profilePicture?: string
  }
}

export interface BookingPayment {
  id: string
  bookingId: string
  
  // Información del pago
  paymentType: PaymentType
  amount: number
  currency: string
  
  // Estado
  status: BookingPaymentStatus
  
  // Stripe
  paymentMethod?: string
  stripePaymentIntentId?: string
  stripeChargeId?: string
  stripeRefundId?: string
  cardLast4?: string
  cardBrand?: string
  
  // Timestamps
  createdAt: string
  updatedAt: string
  completedAt?: string
  refundedAt?: string
  
  // Errores
  errorMessage?: string
  errorCode?: string
  
  metadata?: Record<string, any>
}

export interface BookingCalendar {
  id: string
  listingId: string
  listingCreatedAt: string
  date: string  // ISO date string
  
  // Disponibilidad
  isAvailable: boolean
  
  // Precios dinámicos
  priceOverride?: number
  minimumNights?: number
  
  // Referencia a reserva
  bookingId?: string
  
  notes?: string
  createdAt: string
  updatedAt: string
}

// DTOs para crear/actualizar bookings
export interface CreateBookingDto {
  listingId: string
  checkInDate: string
  checkOutDate: string
  numberOfGuests: number
  guestMessage?: string
}

export interface ConfirmBookingDto {
  bookingId: string
  hostResponse?: string
}

export interface ProcessPaymentDto {
  bookingId: string
  paymentType: PaymentType
  paymentMethodId: string  // Stripe payment method
}

export interface CancelBookingDto {
  bookingId: string
  cancellationReason?: string
}

// Respuestas enriquecidas
export interface BookingWithPayments extends Booking {
  payments: BookingPayment[]
  totalPaid: number
  pendingAmount: number
}

export interface BookingSummary {
  bookingId: string
  totalPrice: number
  expectedReservationPayment: number
  expectedCheckinPayment: number
  totalPaid: number
  reservationPaid: number
  checkinPaid: number
  pendingAmount: number
}

// Para el calendario de disponibilidad
export interface DateAvailability {
  date: string
  isAvailable: boolean
  price?: number
  minimumNights?: number
  bookingId?: string
}

export interface AvailabilityCheckResult {
  available: boolean
  dates: DateAvailability[]
  totalPrice: number
  nights: number
}

// Traducciones de estados
export const BOOKING_STATUS_LABELS: Record<BookingStatus, string> = {
  pending_confirmation: 'Esperando confirmación',
  confirmed: 'Confirmada',
  reservation_paid: 'Reserva pagada (50%)',
  checked_in: 'Check-in realizado',
  completed: 'Completada',
  cancelled_by_guest: 'Cancelada por huésped',
  cancelled_by_host: 'Cancelada por propietario',
  cancelled_no_payment: 'Cancelada por falta de pago',
  refunded: 'Reembolsada'
}

export const PAYMENT_STATUS_LABELS: Record<BookingPaymentStatus, string> = {
  pending: 'Pendiente',
  processing: 'Procesando',
  completed: 'Completado',
  failed: 'Fallido',
  refunded: 'Reembolsado',
  partially_refunded: 'Parcialmente reembolsado'
}

export const PAYMENT_TYPE_LABELS: Record<PaymentType, string> = {
  reservation: 'Pago de reserva (50%)',
  checkin: 'Pago al check-in (50%)',
  full: 'Pago completo (100%)',
  refund: 'Reembolso'
}

// Colores para estados
export const BOOKING_STATUS_COLORS: Record<BookingStatus, string> = {
  pending_confirmation: 'yellow',
  confirmed: 'blue',
  reservation_paid: 'indigo',
  checked_in: 'green',
  completed: 'gray',
  cancelled_by_guest: 'red',
  cancelled_by_host: 'red',
  cancelled_no_payment: 'red',
  refunded: 'orange'
}

File: Backend/app/models/booking.py
"""
Modelos para el sistema de reservas Airbnb
"""
from sqlalchemy import Column, String, Integer, Numeric, Text, TIMESTAMP, ForeignKey, CheckConstraint, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import UUID, JSONB, DATE
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
import enum

from ..core.database import Base


class BookingStatus(str, enum.Enum):
    """Estados de las reservas"""
    PENDING_CONFIRMATION = "pending_confirmation"
    CONFIRMED = "confirmed"
    RESERVATION_PAID = "reservation_paid"
    CHECKED_IN = "checked_in"
    COMPLETED = "completed"
    CANCELLED_BY_GUEST = "cancelled_by_guest"
    CANCELLED_BY_HOST = "cancelled_by_host"
    CANCELLED_NO_PAYMENT = "cancelled_no_payment"
    CANCELLED_PAYMENT_EXPIRED = "cancelled_payment_expired"  # Pago no recibido en 6 horas
    REFUNDED = "refunded"


class BookingPaymentStatus(str, enum.Enum):
    """Estados de los pagos de reserva"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    REFUNDED = "refunded"
    PARTIALLY_REFUNDED = "partially_refunded"


class PaymentType(str, enum.Enum):
    """Tipos de pago"""
    RESERVATION = "reservation"  # 50% inicial
    CHECKIN = "checkin"          # 50% al check-in
    FULL = "full"                # 100% completo
    REFUND = "refund"            # Reembolso


class Booking(Base):
    """Modelo de reservas Airbnb"""
    __tablename__ = "bookings"
    __table_args__ = {"schema": "core"}

    # Identificadores
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    listing_id = Column(UUID(as_uuid=True), nullable=False)
    listing_created_at = Column(TIMESTAMP(timezone=True), nullable=False)
    guest_user_id = Column(UUID(as_uuid=True), nullable=False)
    host_user_id = Column(UUID(as_uuid=True), nullable=False)

    # Fechas de la reserva
    check_in_date = Column(DATE, nullable=False)
    check_out_date = Column(DATE, nullable=False)
    nights = Column(Integer, nullable=False)

    # Información de precios
    price_per_night = Column(Numeric(10, 2), nullable=False)
    total_price = Column(Numeric(10, 2), nullable=False)
    reservation_amount = Column(Numeric(10, 2), nullable=False)  # 50%
    checkin_amount = Column(Numeric(10, 2), nullable=False)      # 50%
    service_fee = Column(Numeric(10, 2), default=0.00)
    cleaning_fee = Column(Numeric(10, 2), default=0.00)

    # Estado
    status = Column(String(50), nullable=False, default='pending_confirmation')

    # Información de huéspedes
    number_of_guests = Column(Integer, nullable=False)

    # Comunicación
    guest_message = Column(Text, nullable=True)
    host_response = Column(Text, nullable=True)
    cancellation_reason = Column(Text, nullable=True)

    # Timestamps
    created_at = Column(TIMESTAMP(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(TIMESTAMP(timezone=True), nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    confirmed_at = Column(TIMESTAMP(timezone=True), nullable=True)
    reservation_paid_at = Column(TIMESTAMP(timezone=True), nullable=True)
    checked_in_at = Column(TIMESTAMP(timezone=True), nullable=True)
    completed_at = Column(TIMESTAMP(timezone=True), nullable=True)
    cancelled_at = Column(TIMESTAMP(timezone=True), nullable=True)
    payment_deadline = Column(TIMESTAMP(timezone=True), nullable=True)  # Plazo de 6h para pagar
    
    # Comprobante de pago
    payment_proof_url = Column(Text, nullable=True)  # URL del comprobante subido
    payment_proof_uploaded_at = Column(TIMESTAMP(timezone=True), nullable=True)
    payment_verified_by = Column(UUID(as_uuid=True), nullable=True)  # Admin que verificó
    payment_verified_at = Column(TIMESTAMP(timezone=True), nullable=True)

    # Metadata (mapeado como extra_metadata para evitar conflicto)
    extra_metadata = Column('metadata', JSONB, default={})


class BookingPayment(Base):
    """Modelo de pagos de reserva"""
    __tablename__ = "booking_payments"
    __table_args__ = {"schema": "core"}

    # Identificadores
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    booking_id = Column(UUID(as_uuid=True), nullable=False)

    # Información del pago
    payment_type = Column(String(20), nullable=False)  # reservation, checkin, full, refund
    amount = Column(Numeric(10, 2), nullable=False)
    status = Column(String(50), nullable=False, default='pending')

    # Integración con Stripe
    stripe_payment_intent_id = Column(String(255), nullable=True)
    stripe_charge_id = Column(String(255), nullable=True)

    # Timestamps
    created_at = Column(TIMESTAMP(timezone=True), nullable=False, default=datetime.utcnow)
    paid_at = Column(TIMESTAMP(timezone=True), nullable=True)

    # Metadata (mapeado como extra_metadata para evitar conflicto)
    extra_metadata = Column('metadata', JSONB, default={})


class BookingCalendar(Base):
    """Modelo de calendario de disponibilidad"""
    __tablename__ = "booking_calendar"
    __table_args__ = {"schema": "core"}

    # Identificadores
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    listing_id = Column(UUID(as_uuid=True), nullable=False)
    date = Column(DATE, nullable=False)

    # Disponibilidad
    is_available = Column(Integer, nullable=False, default=1)  # 1 = disponible, 0 = no disponible
    booking_id = Column(UUID(as_uuid=True), nullable=True)

    # Precio
    price = Column(Numeric(10, 2), nullable=True)
    
    # Notas
    notes = Column(Text, nullable=True)

    # Timestamps
    created_at = Column(TIMESTAMP(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(TIMESTAMP(timezone=True), nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
